UN-OCHA/common_design
# CD-304 Form improvements

## Types of changes
:heavy_check_mark: Improvement (non-breaking change which iterates on an existing feature)

## Motivation and Context
Improves generic styles for `cd-form` component (for Node and Taxonomy term edit forms, User forms, or others) that can be extended or overridden when customisation is necessary.
  
## Impact
For Drupal sites, the [theme_switcher](https://www.drupal.org/project/theme_switcher) module should be installed and configured so the `common_design_subtheme` is used for editing entities.
This results in aesthetic changes and should improve overall consistency and improve editor UX within a specific site and also between sites using the `common_design` base theme.
For developers, there are now template overrides for Node, Taxonomy term, and user edit forms, and various templates for form fields to add classes.

## Description
- Separate `cd-form` CSS to include stylesheets specific to Drupal fields and Media widget but keep in same library definition
- Use styles from RW for generic form field styles with some adjustments
- Add max-width to most input types based on HID styles
- Declare max-width 100% for search form input in CD sass partial, and regenerate CSS
- Add button variant classes for Cancel and Delete buttons via `form alter` function
- Copy theme functions based on work from ReliefWeb relating to User and Taxonomy term form theme hooks https://github.com/UN-OCHA/rwint9-site/pull/163 for consistent template override names, page title display and view mode classes
- Copy Node edit form template override from Stable base theme for customisation ease and to help consistency with the new user and term edit form templates
- Add a variety of form field template overrides (based mainly on work from RW) to add classes, additional markup and attach libraries
- Add User login form template overrides based on work for Sesame site (These forms are not usually used on sites because we redirect to HID for login)
- Add details to `cd-form` README explaining the use of `cd-flow` and how to adjust vertical rhythm as needed

## Steps to test
This is deployed to https://feature.commondesign-unocha-org.ahconu.org and the Basic page content type is configured with many field types as examples.

Since it's form-related, log in [via jenkins](https://jenkins.aws.ahconu.org/view/Common%20Design/job/ds-commondesign-feature-login-url/), visit some entity  edit forms and interact with some fields. Feel free to add content to the fields, and click to delete nodes or terms to hit the confirmation page.
Eg. 
https://feature.commondesign-unocha-org.ahconu.org/node/1/edit
https://feature.commondesign-unocha-org.ahconu.org/taxonomy/term/1/edit
https://feature.commondesign-unocha-org.ahconu.org/user/[ID]/edit

In general, things should appear properly (but minimally) styled and aligned etc.

Also check [Sesame](https://auth.ahconu.org/) for the user login form styles. Log in with your user.
Eg.
https://auth.ahconu.org/user/[ID]
https://auth.ahconu.org/user/login
https://auth.ahconu.org/user/password

**Optionally:**
Most testing of this PR has been done via recent Reliefweb work https://github.com/UN-OCHA/rwint9-site/pulls and is deployed to https://dev.reliefweb-int.ahconu.org (Log in [via jenkins](https://jenkins.aws.ahconu.org/view/ReliefWeb/job/rwint-site-dev-login-url/), visit Node, Term and User form edit pages)
Eg. 
https://dev.reliefweb-int.ahconu.org/node/3758032/edit
https://dev.reliefweb-int.ahconu.org/taxonomy/term/13/edit
https://dev.reliefweb-int.ahconu.org/taxonomy/term/47463/profile

You might notice some visual differences (fields span 100% width, vertical rhythm, widgets) but in general, there should be visual consistency and easy overriding of RW-specific styles where relevant.

## PR Checklist
- [x] I have included a CHANGELOG entry with the PR
- [x] I have run local tests and the tests pass.
- [x] I have linted my code locally.
- [x] **(N/A)** I have made changes to the sub theme to reflect those in the base theme
